common-ts/src/clients/marketDataFeed.ts

```diff
@@ -0,0 +1,710 @@
+
+	// Lookup table for any existing subscriptions which have a compatible configuration for a candle subscriber. Should only be one compatible subscription for a candle subscriber.
+	private compatibleCandleSubscriptionLookup = new CandleSubscriptionLookup();
+
+	/**
+	 * Handles a new subscription by adding it to the necessary lookup tables
+	 * @param subscription
+	 */
+	private handleNewApiSubscription(subscription: ApiSubscription) {
+		this.apiSubscriptions.set(subscription.id, subscription);
+
+		// Add to the trade subscription compatibility lookup
+		this.compatibleTradeSubscriptionLookup.add(subscription);
+
+		// Add to the candle subscription compatibility lookup if a previous one suiting this subscription doesn't already exist
+		if (
+			!this.compatibleCandleSubscriptionLookup.has(
+				subscription.candleSubscriptionsLookupKey
+			)
+		) {
+			this.compatibleCandleSubscriptionLookup.add(subscription);
+		}
+
+		this.checkForTradeSubscriptionTransferForNewSubscription(subscription);
+	}
+
+	private transferTradeSubscribersToNewSubscription(
+		existingTradeSubscription: ApiSubscription,
+		newSubscription: ApiSubscription
+	) {
+		invariant(
+			existingTradeSubscription.id !== newSubscription.id,
+			'Expected different subscriptions when transferring trade subscribers'
+		);
+		invariant(
+			existingTradeSubscription.candleSubscribers.size === 0,
+			'Expected existing trade subscription to have no candle subscribers when transferring trade subscribers'
+		);
+
+		// Transfer the subscribers to the new subscription
+		const tradeSubscribers = Array.from(
+			existingTradeSubscription.tradeSubscribers.values()
+		);
+
+		if (tradeSubscribers.length === 0) return; // Skip early if there are no trade subscribers to transfer
+
+		console.log(
+			`marketDataFeed::transferring_previous_trade_subscribers_to_new_subscription`
+		);
+
+		tradeSubscribers.forEach((tradeSubscriber) => {
+			newSubscription.attachNewTradeSubscriberToExistingSubscription(
+				tradeSubscriber
+			);
+
+			existingTradeSubscription.removeSubscriber(tradeSubscriber.id);
+		});
+	}
+
+	/**
+	 * When we have a new subscription, we want to check if there is an existing subscription with trade subscribers which should be transferred to the new subscription.
+	 *
+	 * Reasoning:
+	 * - If a TRADE SUBSCRIBER caused a new subscription to be created
+	 * - and then a following CANDLE SUBSCRIBER is created for the same market
+	 * => then it is wasteful to keep the previous trade subscription open, because the new candle subscription can collect the data for both subscribers
+	 * @param newSubscription
+	 */
+	private checkForTradeSubscriptionTransferForNewSubscription(
+		newSubscription: ApiSubscription
+	) {
+		const tradeSubscriptionLookupKey =
+			newSubscription.tradeSubscriptionsLookupKey;
+
+		const existingTradeSubscription =
+			this.compatibleTradeSubscriptionLookup.get(tradeSubscriptionLookupKey);
```

Comment:
not obvious to me yet, but this returns the first ApiSubscription in the list, does it not need to check with the rest?